text/0162-refactoring-support.md

```diff
@@ -0,0 +1,606 @@
+case: there's no way to tell them which is which in case they both move to a
+different stack or get renamed.
+
+Indistinguishable resources in this sense are said to be _equivalent_ (see
+Appendix A for a more formal definition). In the unlikely event that two or more
+equivalent resources move, the CLI won't be able to proceed. For example,
+suppose you have two queues in the same stack, named `Queue1` and `Queue2`, with
+the same properties, and without a user defined physical ID:
+
+    App
+    └─ Stack
+       ├─ Queue1
+       └─ Queue2
+
+If they get renamed to, let's say, `Queue3` and `Queue4`,
+
+    App
+    └─ Stack
+       ├─ Queue3
+       └─ Queue4
+
+then the CLI will not be able to establish a 1:1 mapping between the old and new
+names. In this case, it will show you the ambiguity, and stop the deployment:
+
+    Ambiguous Resource Name Changes
+    ┌───┬──────────────────────┐
+    │   │ Resource             │
+    ├───┼──────────────────────┤
+    │ - │ Stack.Queue1A4198146 │
+    │   │ Stack.Queue2B0BA5D32 │
+    ├───┼──────────────────────┤
+    │ + │ Stack.Queue3C7606C37 │
+    │   │ Stack.Queue4D681F510 │
+    └───┴──────────────────────┘
+
+    If you want to take advantage of automatic resource refactoring, avoid 
+    renaming or moving multiple identical resources at the same time.
+
+    If you want to provide an explicit mapping, use the --import-mapping option.
+
+As the message suggests, you can use the `--import-mapping` option as a way to
+resolve ambiguities.
+
+### Programmatic access
+
+The same refactoring feature is also available in the CDK toolkit library:
+
+```typescript
+declare const toolkit: Toolkit;
+declare const cx: ICloudAssemblySource;
+
+// To execute possible refactors as part of the deploy operation:
+await toolkit.deploy(cx, {
+  refactoring: RefactoringMode.EXECUTE_AND_DEPLOY
+});
+
+// Or, if you just want to refactor the stacks:
+await toolkit.refactor(cx);
+```
+
+In case of ambiguity, the toolkit will throw an `AmbiguityError`:
+
+```typescript
+try {
+  await toolkit.refactor(cxSource);
+} catch (e) {
+  if (e instanceof AmbiguityError) {
+    // Handle ambiguity
+  } else {
+    throw e;
+  }
+}
+```
+
+---
+
+
+Ticking the box below indicates that the public API of this RFC has been
+signed-off by the API bar raiser (the `status/api-approved` label was applied to
+the RFC pull request):
+
+```
+
+[ ] Signed-off by API Bar Raiser @rix0rrr
+
+```
+
+## Public FAQ
+
+### What are we launching today?
+
+A new developer experience for CDK users, that allows them to change the
+location of a construct (stack plus logical ID) without causing resource
+replacement. This new experience is available in the CDK CLI `deploy`
+and `refactor` commands, as well as in the toolkit library.
+
+### Why should I use this feature?
+
+If you ever find yourself doing one of the following, you will benefit from
+stack refactoring support:
+
+- Renaming constructs, either intentionally or by mistake.
```

Comment:
What about moving from nearly equivalent constructs, perhaps an org created lambda construct to the cdk lib lambda construct and vice versa (also see my comment asking if we can detect near equivalence, in this case maybe the org version sets a default the user was not aware of, if we could make them aware they could change their code to explicitly match the default or maybe we could allow them to refactor and change the default)